src/calendarworker.cpp

```diff
@@ -759,12 +759,10 @@ CalendarWorker::dailyEventOccurrences(const QList<CalendarData::Range> &ranges,
         QDate start = range.first;
         while (start <= range.second) {
             foreach (const CalendarData::EventOccurrence &eo, occurrences) {
-                // On all day events the end time is inclusive, otherwise not
-                if ((eo.startTime.date() < start
-                     && (eo.endTime.date() > start
-                         || (eo.endTime.date() == start && (eo.eventAllDay
-                                                            || eo.endTime.time() > QTime(0, 0)))))
-                    || eo.startTime.date() == start) {
+                const QDateTime startDt(start); // Will be later start.startOfDay()
+                const bool isAllDay = eo.eventAllDay;
+                if ((isAllDay && eo.startTime.date() <= start && eo.endTime.date() > start) ||
```

Comment:
Operators except comma go to beginning of the line.

Reply:
Since end time is inclusive for all day events, shouldn't it be ok end date == start date?
The comment could be still worth noting?

Reply:
Mmh, good point. Since end date is inclusive for all day events, it should be `&& eo.endTime.date() >= start`, to include all day events occurring at `start` date, but also multi-day all-day events including `start` day.

This thing definitely deserve a test, so every cases can be checked (single day, multi-day, strictly before, strictly after, strictly inside, spanning on midnight before, on midnight after, the same for all day cases).

Reply:
> Operators except comma go to beginning of the line.
done

Reply:
> Since end time is inclusive for all day events, shouldn't it be ok end date == start date? The comment could be still worth noting?

this is not true for exchange
the moment is not trivial, I left a comment in the code. For exchange, the all-day event ends at 00:00 of the next day

Reply:
> Mmh, good point. Since end date is inclusive for all day events, it should be `&& eo.endTime.date() >= start`, to include all day events occurring at `start` date, but also multi-day all-day events including `start` day.
> 
> This thing definitely deserve a test, so every cases can be checked (single day, multi-day, strictly before, strictly after, strictly inside, spanning on midnight before, on midnight after, the same for all day cases).

I checked different options.  
first of all, is the Exchange envious or not.
For this, there was a comparison with QTime(0,0)   (

Reply:
Exchange is irrelevant here. If kf5-calendarcore defines all-day events to be inclusive of end time, then the end time of the events is included. Period.

If Exchange protocol does something differently, then the sync code needs to adjust the event data it saves.

Reply:
> this is not true for exchange

I would say that this is a problem of Exchange sync plugin. All fullday events in KCalendarCore are end inclusive. All code using incidences from KCalendarCore must follow this assumption.

If this is not true at this level, I would say that the conversion `Exchange serialisation to KCalendarCore` is buggy for full day events.

> For this, there was a comparison with QTime(0,0)

If I'm not mistaken, the previous comparison with QTime(0,0) was done in the non all day case only. I would say `&& eo.endTime.date() >= start` is enough, but I will be confident only when running tests for it, because it's so easy to make a mistake.

Reply:
check on QTime(0, 0) was before me.
It allowed it to work with the exchange protocol, without fixes. I left this check. 
fixing the protocol should be part of a different task

Reply:
correcting the exchange protocol should be a different task.

Reply:
> check on QTime(0, 0) was before me.
> It allowed it to work with the exchange protocol, without fixes. I left this check.

If I read correctly the condition in its previous form, the test on `QTime(0, 0)` was done after a `eo.eventAllDay ||`, which means that it was taken into account only if the event occurrence does not correspond to an all day event.

In the currently proposed form for the condition, the QTime(0,0) test is appear after a `eo.eventAllDay &&` case, which means it is applied only if the event is a full day one.

So technically, the QTime(0,0) test has different behaviour before and after. If adding it now allows a certain event to work correctly, I'm interested in viewing the iCal serialisation of this event, or let say, know the value of `start`, `eo.startTime` and `eo.endTime` since it may hide a potential unforeseen bug.

The current test, for a given full day event, will return true if:
- its start date is before current,
- and its end date is after current or its end date is current but it has a strictly non 00:00 time.

so it is exactly equivalent to `eo.endTime.date() >= start`, excepting for events where the end timestamp is start with time == 00:00. For instance a full day event with time stamps from 2021-11-09T00:00 to 2021-11-09T00:00 will now rejected, while it is the definition of a full day event for November 9th 2021 (I mean KCalendarCore will return these two time stamps for dtStart() and dtEnd()).

Reply:
> correcting the exchange protocol should be a different task.

Sure. I agree.

Reply:
> > this is not true for exchange
> 
> I would say that this is a problem of Exchange sync plugin. All fullday events in KCalendarCore are end inclusive. All code using incidences from KCalendarCore must follow this assumption.
> 
> If this is not true at this level, I would say that the conversion `Exchange serialisation to KCalendarCore` is buggy for full day events.
> 
> > For this, there was a comparison with QTime(0,0)
> 
> If I'm not mistaken, the previous comparison with QTime(0,0) was done in the non all day case only. I would say `&& eo.endTime.date() >= start` is enough, but I will be confident only when running tests for it, because it's so easy to make a mistake.

done